File: DataStructure/BST/merge_bst.py
""" Merge 2 BSTs with limited extra space """

class Node:
    def __init__(self,key):
        self.key = key
        self.left = self.right = None

class Stack:

    def __init__(self):
        self.data = list()

    def push(self, val):
        self.data.append(val)

    def pop(self):
        if not len(self):
            raise Exception('Empty Stack')
        return self.data.pop()

    def __len__(self):
        return len(self.data)

    def top(self):
        if not len(self):
            raise Exception('Empty Stack')
        return self.data[-1]

    def __bool__(self):
        return bool(self.data)

def inorder(root):
    if root:
        inorder(root.left)
        print(root.key, end=' ')
        inorder(root.right)

def merge_bst(root1, root2):
    if not root1:
        inorder(root2)
    elif not root2:
        inorder(root1)

    stack1 = Stack()
    stack2 = Stack()

    curr1 = root1
    curr2 = root2

    while curr1 or curr2 or stack1 or stack2:
        if curr1:
            while curr1:
                stack1.push(curr1)
                curr1 = curr1.left

        if curr2:
            while curr2:
                stack2.push(curr2)
                curr2 = curr2.left

        if not (stack1 and stack2):
            if stack1:
                while stack1:
                    temp1 = stack1.pop()
                    temp1.left = None
                inorder(temp1)
                break

            elif stack2:
                while stack2:
                    temp2 = stack2.pop()
                    temp2.left = None
                inorder(temp2)
                break

        temp1 = stack1.top()
        temp2 = stack2.top()

        if temp1.key < temp2.key:
            temp1 = stack1.pop()
            print(temp1.key, end=' ')
            curr1 = temp1.right

        elif temp2.key < temp1.key:
            temp2 = stack2.pop()
            print(temp2.key, end=' ')
            curr2 = temp2.right

if __name__ == '__main__':
    root1 = Node(3)
    root1.left = Node(1)
    root1.right = Node(5)
  
    """ Let us create the following tree as second tree  
            4  
        / \  
        2 6  
    */
    """
    root2 = Node(4);  
    root2.left = Node(2);  
    root2.right = Node(6);  
  
    merge_bst(root1, root2);  

File: LeetCode/Google/merge_bst.cpp
// https://www.geeksforgeeks.org/merge-two-bsts-with-limited-extra-space/
#include <bits/stdc++.h>
using namespace std;

struct Node{
	int data;
	Node* left;
	Node* right;
	Node(int data = 0): data{data}, left{nullptr}, right{nullptr}{}
};

void merge(Node* root1, Node* root2){
	stack<Node*> st1;
	stack<Node*> st2;
	Node *curr1 = root1, *curr2 = root2;
	while(true){
		while(curr1){
			st1.push(curr1);
			curr1 = curr1 -> left;
		}
		while(curr2){
			st2.push(curr2);
			curr2 = curr2 -> left;
		}
		if(!st1.empty() && !st2.empty()){
			Node* top1 = st1.top();
			Node* top2 = st2.top();

			if(top1->data < top2->data){
				cout << top1->data << " ";
				curr1 = top1->right; st1.pop();
			}
			else{
				cout << top2->data << " ";
				curr2 = top2->right; st2.pop();
			}
		}
		else{
			if(!st1.empty()){
				cout << (st1.top())->data << " ";
				curr1 = st1.top()->right; st1.pop();
			}
			else if(!st2.empty()){
				cout << (st2.top())->data << " ";
				curr2 = st2.top()->right; st2.pop();
			}
			else return;
		}
	}
}

int main(){
	Node *root1 = NULL, *root2 = NULL;

    /* Let us create the following tree as first tree
            3
        / \
        1 5
    */
    root1 = new Node(3);
    root1->left = new Node(1);
    root1->right = new Node(5);

    /* Let us create the following tree as second tree
            4
        / \
        2 6
    */
    root2 = new Node(4);
    root2->left = new Node(2);
    root2->right = new Node(6);

    // Print sorted nodes of both trees
    merge(root1, root2);
}
